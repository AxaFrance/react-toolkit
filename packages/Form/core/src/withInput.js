import { ClassManager } from '@axa-fr/react-toolkit-core';
import Constants from './InputConstants';
import React from "react";

export const omit = keys => props => {
  if (!keys) {
    return props;
  }
  const clonedProps = { ...props };
  keys.forEach(element => {
    if (element in clonedProps) {
      delete clonedProps[element];
    }
  });
  return clonedProps;
};

const defaultOnChange = ({ name, onChange }) => e =>
  onChange({ value: e.target.value, name, id: e.target.id });

export const withInput = (
  defaultClassName,
  addPropTypes = {},
  addDefaultProps = {},
  withHandlersOverride = {},
  withPropsOverride = null
) => Component =>  {
  let defaultWithProps = ({ className, classModifier }) => ({
    componentClassName: ClassManager.getComponentClassName(
      className,
      classModifier,
      defaultClassName
    ),
  });
  if (withPropsOverride) {
    defaultWithProps = withPropsOverride;
  }

  const handlers = {
    onChange: defaultOnChange,
    ...withHandlersOverride,
  };

  const NewComponent = (props) => {
    const {isVisible} = props;
    if (!isVisible) {
      return null;
    }
    const onHandlers = {};
    for (let propertyName in handlers){
      onHandlers[propertyName] = handlers[propertyName](props);
    }
    return <Component {...props} {...defaultWithProps(props)} {...onHandlers} />;
  };

  const propTypes = {
    ...Constants.propTypes,
    ...addPropTypes,
  };
  NewComponent.propTypes = propTypes;

  const defaultProps = {
    ...Constants.defaultProps,
    ...addDefaultProps,
  };
  NewComponent.defaultProps = defaultProps;

 return NewComponent;
};
