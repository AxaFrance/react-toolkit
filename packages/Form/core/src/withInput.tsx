import React, { ChangeEvent, ComponentType } from 'react';
import { ClassManager } from '@axa-fr/react-toolkit-core';
import Constants from './InputConstants';
import { CustomFormEvent } from './customEvent';

export const omit =
  (keys?: string[]) =>
  <P extends {}>(props: P) => {
    if (!keys) {
      return props;
    }

    return Object.entries(props).reduce(
      (acc, [key, value]) =>
        !keys.includes(key)
          ? {
              ...acc,
              [key]: value,
            }
          : acc,
      {}
    );
  };

type DefaultOnChangeProps = {
  name: string;
  onChange: (event: CustomFormEvent) => void;
};

export const defaultOnChange =
  ({ name, onChange }: DefaultOnChangeProps) =>
  (
    e: ChangeEvent<HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement>
  ) =>
    onChange({ name, value: e.target.value, id: e.target.id });

const defaultWithProps =
  (defaultClassName: string) =>
  ({
    className,
    classModifier,
  }: {
    className?: string;
    classModifier?: string;
  }) => ({
    componentClassName: ClassManager.getComponentClassName(
      className,
      classModifier,
      defaultClassName
    ),
  });

const setOnHandlers = (handlers: any = {}, props: any) =>
  Object.entries(handlers).reduce(
    (acc, [propertyName]) => ({
      ...acc,
      [propertyName]: handlers[propertyName](props),
    }),
    {}
  );

export const withInput =
  <
    P extends {
      className?: string;
      classModifier?: string;
      isVisible?: boolean;
    }
  >(
    defaultClassName: string,
    addPropTypes = {},
    addDefaultProps = {},
    withHandlersOverride = {},
    withPropsOverride = defaultWithProps(defaultClassName),
    defaultOnChangeFn = defaultOnChange
  ) =>
  (Component: ComponentType<P>) => {
    const EnhancedInput = (props: P) => {
      const { isVisible } = props;
      if (!isVisible) {
        return null;
      }

      return (
        <Component
          {...props}
          {...withPropsOverride(props)}
          {...setOnHandlers(
            {
              onChange: defaultOnChangeFn,
              ...withHandlersOverride,
            },
            props
          )}
        />
      );
    };

    EnhancedInput.propTypes = {
      ...Constants.propTypes,
      ...addPropTypes,
    };

    EnhancedInput.defaultProps = {
      ...Constants.defaultProps,
      ...addDefaultProps,
    };

    return EnhancedInput;
  };
