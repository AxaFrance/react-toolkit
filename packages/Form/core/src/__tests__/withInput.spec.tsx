import React from 'react';
import { render } from '@testing-library/react';
import { omit, defaultOnChange, withInput } from '../withInput';

describe('omit', () => {
  it('Should return only wanted props when omit have been called', () => {
    const newProps = omit(['classModifier', 'isVisible'])({
      classModifier: 'modifier',
      className: 'classname',
      isVisible: true,
    });
    expect(newProps).toEqual({ className: 'classname' });
  });

  it('Should return the same props when omit have been called with no key', () => {
    const newProps = omit()({
      classModifier: 'modifier',
      className: 'classname',
    });
    expect(newProps).toEqual({
      classModifier: 'modifier',
      className: 'classname',
    });
  });
});

describe('defaultOnChange', () => {
  const onChangeMock = jest.fn();
  it('Should onChange to be called with name, target.id and target.value when defaultOnChange have been called with name and event', () => {
    defaultOnChange({ name: 'myFieldName', onChange: onChangeMock })({
      target: {
        id: 'id',
        value: 'value',
      },
    } as React.ChangeEvent<HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement>);
    expect(onChangeMock).toHaveBeenCalledWith({
      name: 'myFieldName',
      id: 'id',
      value: 'value',
    });
  });
});

describe('withInput', () => {
  const onChangeMock = (e: any) => console.log(e);
  const Input = (props: any) => (
    <>
      <p>{props.componentClassName}</p>
      <p>{props.classModifier}</p>
      <input type="text" name={props.name} id="name" />
    </>
  );
  const defaultProps = {
    name: 'name',
    onChange: onChangeMock,
  };
  it('Should return Enhanced Input when withInput have been called with default className and input component', () => {
    const NewInput = withInput('default')(Input);
    const { asFragment, getByText } = render(<NewInput {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
    expect(getByText('default')).toBeInTheDocument();
  });

  it('Should return Enhanced Input when withInput have been called with default className and input component and modifier', () => {
    const NewInput = withInput('default')(Input);
    const { asFragment, getByText } = render(
      <NewInput {...defaultProps} classModifier="custom" />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByText('default default--custom')).toBeInTheDocument();
    expect(getByText('custom')).toBeInTheDocument();
  });

  it('Should not return Enhanced Input when isVisible is false', () => {
    const NewInput = withInput('default')(Input);
    const { asFragment, queryByText } = render(
      <NewInput {...defaultProps} isVisible={false} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(queryByText('default default--custom')).not.toBeInTheDocument();
    expect(queryByText('custom')).not.toBeInTheDocument();
  });
});
