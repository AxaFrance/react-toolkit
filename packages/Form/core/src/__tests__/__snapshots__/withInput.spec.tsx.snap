// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withInput Should not return Enhanced Input when isVisible is false 1`] = `<DocumentFragment />`;

exports[`withInput Should return Enhanced Input when withInput have been called with default className and input component 1`] = `
<DocumentFragment>
  <p>
    default
  </p>
  <p />
  <input
    id="name"
    name="name"
    type="text"
  />
</DocumentFragment>
`;

exports[`withInput Should return Enhanced Input when withInput have been called with default className and input component and modifier 1`] = `
<DocumentFragment>
  <p>
    default default--custom
  </p>
  <p>
    custom
  </p>
  <input
    id="name"
    name="name"
    type="text"
  />
</DocumentFragment>
`;
